// os-browserify/browser.js
exports.endianness = function () { return 'LE' ***REMOVED***

exports.hostname = function () {
    if (typeof location !== 'undefined') {
        return location.hostname
    }
    else return '';
***REMOVED***

exports.loadavg = function () { return [] ***REMOVED***

exports.uptime = function () { return 0 ***REMOVED***

exports.freemem = function () {
    return Number.MAX_VALUE;
***REMOVED***

exports.totalmem = function () {
    return Number.MAX_VALUE;
***REMOVED***

exports.cpus = function () { return [] ***REMOVED***

exports.type = function () { return 'Browser' ***REMOVED***

exports.release = function () {
    if (typeof navigator !== 'undefined') {
        return navigator.appVersion;
    }
    return '';
***REMOVED***

exports.networkInterfaces
= exports.getNetworkInterfaces
= function () { return {} ***REMOVED***

exports.arch = function () { return 'javascript' ***REMOVED***

exports.platform = function () { return 'browser' ***REMOVED***

exports.tmpdir = exports.tmpDir = function () {
    return '/tmp';
***REMOVED***

exports.EOL = '\n';

exports.homedir = function () {
	return '/'
***REMOVED***
